### DISCUSSION TITLE
Recommended alternative for json() marked deprecated in 6.28

### DISCUSSION THREAD

**Discussion by RoBackman** [remix-run/react-router#12257] [Q&A]
Hi!

When using future.v7_skipActionErrorRevalidation the json() method has the nice utility of altering response status to 4xx/5xx without the need to explicitly create the response yourself (providing the ability to opt-out of revalidation in case of an error submitting some data etc.).

It also provides a convenient way to require returned data to conform to a given type.

The alternatives i see for typing is probably using "satisfies".
But with that and the revalidation opt-out i see myself recreating the json() method myself, and therefore would be interested to hear the reason why it is beeing deprecated.

**Comment by ingmarh** [comment]
In React Router v7 there is a new `data` utility that can be used to return custom status codes and headers in loaders/actions — [see this section in the WIP v7 docs](https://reactrouter.com/dev/start/framework/data-loading#custom-status-codes-and-headers) (*update*: linked section since removed with an update, but `data` still exists). I was able to use it as a replacement for `json` in a project using the current v7 prerelease. I’m guessing that the `data` utility will become the recommended alternative.

When no custom status codes or headers are needed, [raw objects can be returned from loaders/actions](https://github.com/remix-run/react-router/blob/dd323b6dec531811ce80595ac0521455b46f0bc2/CHANGELOG.md#v6280:~:text=in%20favor%20of%20returning%20raw%20objects).

**Comment by ingmarh** [comment]
There’s also [`Response.json`](https://developer.mozilla.org/en-US/docs/Web/API/Response/json_static) which is [linked in the v7.0.0 changelog](https://github.com/remix-run/react-router/blob/f62986c744b4d06ef318eea83b56bd74bd99f931/CHANGELOG.md#:~:text=You%20can%20use%20Response.json%20if%20you%20still%20need%20to%20construct%20JSON%20responses%20in%20your%20app) as an alternative to `json`.

**Reply by hyusetiawan** [reply]
this does not type the useLoaderData correctly. The only way to get the data to be typed correctly it seems to straight return:
```

export const loader = async ({ request }: LoaderFunctionArgs) => {
  const session = await auth.api.getSession({
    headers: request.headers
  })

  // wrapping this with Response.json or the deprecated json does not type useLoaderData correctly
  return {
    session
  }
}
```
`  const { session } = useLoaderData<typeof loader>()`

**Comment by skizzo** [comment]
So how can I return a loader response without using `json()` but `Response.json()` in case I need to return custom status codes,  headers, etc. which returns correctly typed data via `useLoaderData()`?

**Reply by tomsoderlund** [reply]
@skizzo Did you find a solution here?

**Reply by skizzo** [reply]
Nope, I have to define a type manually and use it for both the return object of my loader and useLoader() to ensure type safety.

**Reply by xeroneon** [reply]
@skizzo You are able to use "data" instead of "json", if you use "data" you will be able to pass status codes and set headers and such as if it was still "json". you will also retain all type safety as long as for the loader and action you only give types to the parameters, if you give an explicit type to the function itself or an explicit type for the return then it generally breaks the type inference

**Reply by madfcat** [reply]
Why don't you guys use `remix-typedjson`? I used the tool and it worked nicely.
I got rid of `JsonifyObject` and could attach status codes to my returns from loader.

**Comment by wolfgang-hartl** [comment]
Hey there, 
following up that topic. I am not really able to get typed actionData in my Page. Using an named export called action, which should return me a data object. Neither Response.json(), returning the raw object, nor using the data utility helps me get that types into my page. actionData is always of type undefined. 

What am i doing wrong? 

I've put together a relative simple demo example: 
```js
export async function action({ request }: Route.ActionArgs) {
  const body = await request.json();

  const resp = await AuthService.postLogin({
    body: {
      email: body.email,
      password: body.password,
    },
  });

  if (resp.response.ok) {
    throw redirect("/");
  }

  return Response.json(resp); // or returning the raw object, or using data() 
}

export default function Home({ actionData }: Route.ComponentProps) {
  const submit = useSubmit();

  return (
    <>
      <Form
        onSubmit={(data) => {
          submit(data, { method: "POST", encType: "application/json" });
        }}
      >
        <TextInput name="email" label="Email" type="email" />
        <TextInput name="password" label="Password" type="password" />

        <button type="submit">Submit</button>
      </Form>

      {actionData && <p className="text-red-500">{actionData.error.message}</p>}
    </>
  );
}
```

**Reply by xeroneon** [reply]
I'm running into the same issue, lsp doesn't pick up the types returned from the action

**Reply by adebiyial** [reply]
@wolfgang-hartl I figured that sometimes, all you need to do is restart your server or reload your code editor.

**Comment by lhbelfanti** [comment]
Hi there!

Checking [this commit](https://github.com/remix-run/react-router/pull/12146/files#diff-83ada76eefb539ee6116b7b31b717768a35bdd12d44e1389ea43a72020a86fe3R63) and reading what @ingmarh and @xeroneon said, I was able to replace the `json()` function with the `data()` function (`import { data } from "@remix-run/node"`). It accepts 2 parameters the `custom type` and the `response type`. The `custom type` can be obtained then using the `useActionData` hook.

Example:
```typescript
export const loader = async ({ request }: LoaderFunctionArgs) => {
  const session = await auth.api.getSession()
  const myCustomType: MyCustomType = {
    isLoggedIn: true,
    token: session.token,
  };

  return data( myCustomType, { status: 200, headers: session.headers })
}
```
 
and then using the `useActionData` hook:
```typescript
  const actionData = useActionData<MyCustomType>();
  console.log(actionData.isLoggedIn);
  console.log(actionData.token);
```

I hope this helps someone who comes across the same problem and finds this thread.

**Reply by tomsoderlund** [reply]
Using `data(..., { status: 400 })` doesn't seem to modify the HTTP status code in the response.

**Comment by tomsoderlund** [comment]
Hey y'all, `data()` is kinda like `json()` **except it doesn't work at all**! It doesn't return status codes or modify headers. Sort of loses the point, right?

**Reply by patrikdevlin** [reply]
In some scenarios where you aren't triggering a loader or action in response to a page / form, you will need to create the Response yourself. E.x if your app is responding to webhooks or an api route outside of the scope of the UI.

Here is a utility function you can use, but it would be nice if this was handled for you.

```typescript
export function json<T>(data: T, init?: ResponseInit): Response {
	return new Response(JSON.stringify(data), {
		status: init?.status || 200,
		statusText: init?.statusText,
		headers: {
			'Content-Type': 'application/json',
			...(init?.headers ?? {}),
		},
	})
}

**Reply by GabenGar** [reply]
This is not correct, `data()` does all of this, the caveat is you have to pass headers in the route module's [`headers()`](https://reactrouter.com/start/framework/route-module#headers) export.

**Comment by renatodex** [comment]
Hey hey, any insights on this? I was returning raw objects but it would be nice to find a good replacement for json.

**Comment by pietmichal** [comment]
Setting status code using e.g. `data(..., { status: 404 })`  doesn't change the returned status code **in actions**.

It works for loaders.
